app/models/concerns/user_omniauth_concern.rb

```diff
@@ -0,0 +1,34 @@
+module UserOmniauthConcern
+  extend ActiveSupport::Concern
+  include UserOmniauthFacebookConcern
+
+  included do
+    devise :omniauthable
+  end
+
+  module ClassMethods
+    # Find or create a user from omniauth data.
+    #
+    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+    # @return [User] The user found or created from omniauth data.
+    def from_omniauth(auth)
+      identity = User::Identity.find_by(provider: auth.provider, uid: auth.uid)
+      return identity.user if identity
+
+      User.create do |user|
+        user.assign_attributes(name: auth.info.name, email: auth.info.email,
```

Comment:
That why I have `if user.persisted`, if not persisted means the information is not enough. Then he will be redirected and required to fill in the missing information himself.